org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/operation/internal/DefaultToolingApiOperationManager.java

```diff
@@ -38,7 +38,7 @@ public void run(ToolingApiOperation runnable, IProgressMonitor monitor) throws C
     @Override
     public void run(ToolingApiOperation runnable, CancellationTokenSource tokenSource, IProgressMonitor monitor) throws CoreException {
         IProgressMonitor efficientMonitor = new RateLimitingProgressMonitor(monitor, 500, TimeUnit.MILLISECONDS);
-        ResourcesPlugin.getWorkspace().run(new WorkspaceRunnableAdapter(runnable, tokenSource), efficientMonitor);
+        new WorkspaceRunnableAdapter(runnable, tokenSource).run(efficientMonitor);
```

Comment:
❌ This makes this use case work incidentally. The proper approach would be to define `ToolingApiOperation#getRule()` that returns an `ISchedulingRule` instance. That rule can be passed to `IWorkspace#run()` and to `ToolingApiJob#setRule()`as well.

Reply:
thanks donat